geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderAndCQDurableClientDUnitTest.java

```diff
@@ -0,0 +1,470 @@
+    }
+  }
+
+  /**
+   * This hook will run some transaction traffic after server requested initial image from
+   * primary bucket server, and then it will check if events are enqueued in temporary queue.
+   * It is expected that events are enqueued in this case because bucket is not yet available.
+   */
+  void configureHooksOnRunningMember(MemberVM server, int bucketId) {
+    server.invoke(() -> InitialImageOperation.setGIITestHook(new InitialImageOperation.GIITestHook(
+        InitialImageOperation.GIITestHookType.AfterSentRequestImage, "_B__test1_" + bucketId) {
+      private static final long serialVersionUID = -3790198435185240444L;
+
+      @Override
+      public void reset() {}
+
+      @Override
+      public void run() {
+        doPutsInServer();
+        ParallelGatewaySenderAndCQDurableClientDUnitTest.IS_HOOK_TRIGGERED = true;
+        if (sizeOfBucketToTempQueueMap("ln") != 0) {
+          ParallelGatewaySenderAndCQDurableClientDUnitTest.IS_TEMP_QUEUE_USED = true;
+        }
+      }
+    }));
+  }
+
+  void resetHooks(MemberVM server) {
+    server.invoke(() -> InitialImageOperation.resetAllGIITestHooks());
+  }
+
+  public static boolean isHookTriggered() {
+    return ParallelGatewaySenderAndCQDurableClientDUnitTest.IS_HOOK_TRIGGERED;
+  }
+
+  public static boolean isTempQueueUsed() {
+    return ParallelGatewaySenderAndCQDurableClientDUnitTest.IS_TEMP_QUEUE_USED;
+  }
+
+  MemberVM getServerToStop(List<MemberVM> list) {
+    for (MemberVM member : list) {
+      if (!member.invoke(this::getPrimaryBucketList).isEmpty()) {
+        return member;
+      }
+    }
+    return null;
+  }
+
+  private void startDurableClient()
+      throws Exception {
+    int locatorPort = locatorSite2.getPort();
+    clientSite3SubscriptionQueue = clusterStartupRule.startClientVM(7, ccf -> ccf
+        .withPoolSubscription(true).withLocatorConnection(locatorPort).withCacheSetup(c -> c
+            .set("durable-client-id", DURABLE_CLIENT_ID)));
+  }
+
+  private void createDurableCQs(String... queries) {
+    clientSite3SubscriptionQueue.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      QueryService queryService = ClusterStartupRule.getClientCache().getQueryService();
+      CqAttributesFactory cqAttributesFactory = new CqAttributesFactory();
+      ParallelGatewaySenderAndCQDurableClientDUnitTest.cqListener =
+          new CqListenerTestReceivedEvents();
+      cqAttributesFactory.addCqListener(cqListener);
+
+      for (String query : queries) {
+        CqQuery cq = queryService.newCq(query, cqAttributesFactory.create(), true);
+        cq.execute();
+      }
+      ClusterStartupRule.getClientCache().readyForEvents();
+    });
+  }
+
+  private Set<Integer> getPrimaryBucketList() {
+    PartitionedRegion region = (PartitionedRegion) Objects
+        .requireNonNull(ClusterStartupRule.getCache()).getRegion("test1");
+    return new TreeSet<>(region.getDataStore().getAllLocalPrimaryBucketIds());
```

Comment:
Does this need to be a `TreeSet`?

Reply:
No, I will change it to HashSet. Is that OK?

Reply:
Yeah, that's fine, thanks